src/KeyboardEvents.cpp

```diff
@@ -0,0 +1,149 @@
+/* a lot of this code was stolen from https://github.com/kernc/logkeys but then modified */
+
+#include "KeyboardEvents.h"
+
+#include <cwctype>
+#include <iostream>
+#include <string>
+
+#include <linux/input.h>
+
+#include "Exceptions.h"
+#include "KeyTable.h"
+#include "Utils.h"
+
+using namespace hemiola;
+
+hemiola::KeyboardEvents::KeyboardEvents ( std::shared_ptr<KeyTable> keyTable )
+    : m_KeyTable ( std::move ( keyTable ) )
+    , m_Device ( std::make_shared<Device>() )
+{}
+
+hemiola::KeyboardEvents::KeyboardEvents ( std::shared_ptr<KeyTable> keyTable, std::shared_ptr<Device> device )
+    : m_KeyTable ( std::move ( keyTable ) )
+    , m_Device ( std::move ( device ) )
+{}
+
+void hemiola::KeyboardEvents::capture (
+    std::function<void ( std::variant<wchar_t, unsigned short> )> onEvent,
+    std::function<void ( std::exception_ptr )> onError )
+{
+    try {
+        m_Device->open();
+        while ( updateKeyState() ) {
+            captureEvent ( onEvent );
+        }
+    } catch ( ... ) {
+        std::cerr << "Input device closed unexpectedly.\n";
+        m_Device->close();
+        onError ( std::current_exception() );
```

Comment:
Might be easier to catch `std::exception` here rather than `...`.  Then have a `catch(...){assert(0)}` to sort out any exceptions that aren't derived from std::exception, IMO that shouldn't happen.

Reply:
Catching `std::exception` will cast the exception to that type instead of the original type.